browser/brave_vpn/dns/brave_vpn_dns_observer_service_unittest.cc

```diff
@@ -0,0 +1,317 @@
+    SystemNetworkContextManager::set_stub_resolver_config_reader_for_testing(
+        stub_resolver_config_reader_.get());
+    dns_service_->SetPrefServiceForTesting(pref_service());
+    dns_service_->SetVPNNotificationCallbackForTesting(base::DoNothing());
+  }
+
+  void TearDown() override {
+    // BraveVpnDnsObserverService destructor must be called before the task
+    // runner is destroyed.
+    dns_service_.reset();
+  }
+  PrefService* local_state() { return &local_state_; }
+  PrefService* pref_service() { return &profile_pref_service_; }
+
+  void FireBraveVPNStateChange(mojom::ConnectionState state) {
+    dns_service_->OnConnectionStateChanged(state);
+  }
+
+  bool WasVpnNotificationShownForMode(const std::string& mode,
+                                      const std::string& servers) {
+    bool callback_called = false;
+    dns_service_->SetVPNNotificationCallbackForTesting(
+        base::BindLambdaForTesting([&]() { callback_called = true; }));
+    SetDNSMode(mode, servers);
+    return callback_called;
+  }
+
+  void SetDNSMode(const std::string& mode, const std::string& doh_providers) {
+    local_state()->SetString(::prefs::kDnsOverHttpsTemplates, doh_providers);
+    local_state()->SetString(::prefs::kDnsOverHttpsMode, mode);
+    SystemNetworkContextManager::GetStubResolverConfigReader()
+        ->UpdateNetworkService(false);
+  }
+
+  bool WasPolicyNotificationShownForState(mojom::ConnectionState state) {
+    bool callback_called = false;
+    dns_service_->SetPolicyNotificationCallbackForTesting(
+        base::BindLambdaForTesting([&]() { callback_called = true; }));
+    FireBraveVPNStateChange(state);
+    return callback_called;
+  }
+
+  void ExpectDNSMode(const std::string& mode,
+                     const std::string& doh_providers) {
+    auto dns_config = SystemNetworkContextManager::GetStubResolverConfigReader()
+                          ->GetSecureDnsConfiguration(false);
+    auto* current_mode = SecureDnsConfig::ModeToString(dns_config.mode());
+    auto current_servers = dns_config.doh_servers().ToString();
+    EXPECT_EQ(current_mode, mode);
+    EXPECT_EQ(current_servers, doh_providers);
+  }
+  void SetManagedMode(const std::string& value) {
+    local_state_.SetManagedPref(::prefs::kDnsOverHttpsMode,
+                                std::make_unique<base::Value>(value));
+  }
+
+ private:
+  std::unordered_map<std::string, std::string> policy_map_;
+  content::BrowserTaskEnvironment task_environment_;
+  std::unique_ptr<BraveVpnDnsObserverService> dns_service_;
+  sync_preferences::TestingPrefServiceSyncable profile_pref_service_;
+  TestingPrefServiceSimple local_state_;
+  std::unique_ptr<StubResolverConfigReader> stub_resolver_config_reader_;
+};
+
+TEST_F(BraveVpnDnsObserverServiceUnitTest, AutoEnable) {
+  // Browser DoH mode off -> override browser config and enable vpn
+  local_state()->ClearPref(::prefs::kBraveVpnDnsConfig);
```

Comment:
why are we clearing this? The default setting is empty so we should just be verifying that as well

Reply:
see comment here https://github.com/brave/brave-core/pull/15989#discussion_r1042003145

Reply:
yes, but that doesn't change the fact that we need to verify the default state somewhere. Testing the defaults was clearly missed in other places like https://github.com/brave/brave-core/pull/15989/files#r1045904676 and we should not make that mistake again here. We should always verify the initial state before changing it